jbmc/src/java_bytecode/convert_java_nondet.cpp

```diff
@@ -29,83 +29,118 @@ Author: Reuben Thomas, reuben.thomas@diffblue.com
-  const auto next_instr=std::next(target);
-  if(!target->is_assign())
-  {
-    return next_instr;
-  }
+  const auto next_instr = std::next(target);
+  bool changed = false;
 
-  // Return if the rhs of the assignment isn't a side effect expression
-  const auto &assign=to_code_assign(target->code);
-  if(assign.rhs().id()!=ID_side_effect)
+  // We only expect to find nondets in the rhs of an assignments or in return
+  // statements
+  for(exprt &op : target->code.operands())
   {
-    return next_instr;
-  }
+    if(op.id() != ID_side_effect)
+    {
+      continue;
+    }
 
-  // Return if the rhs isn't nondet
-  const auto &side_effect=to_side_effect_expr(assign.rhs());
-  if(side_effect.get_statement()!=ID_nondet)
-  {
-    return next_instr;
-  }
+    const auto &side_effect = to_side_effect_expr(op);
+    if(side_effect.get_statement() != ID_nondet)
+    {
+      continue;
+    }
 
-  const auto lhs=assign.lhs();
-  // If the lhs type doesn't have a subtype then I guess it's primitive and
-  // we want to bail out now
-  if(!lhs.type().has_subtype())
-  {
-    return next_instr;
-  }
+    const typet &type = side_effect.type();
+    // If the lhs type doesn't have a subtype then I guess it's primitive and
+    // we want to bail out now
+    if(type.id() != ID_pointer)
+    {
+      continue;
+    }
 
-  // Although, if the type is a ptr-to-void then we also want to bail
-  if(lhs.type().subtype().id()==ID_empty ||
-     lhs.type().subtype().id()==ID_code)
-  {
-    return next_instr;
+    // Although, if the type is a ptr-to-void or a function pointer then we also
+    // want to bail
+    if(type.subtype().id() == ID_empty || type.subtype().id() == ID_code)
+    {
+      continue;
+    }
+
+    // Check whether the nondet object may be null
+    if(!to_side_effect_expr_nondet(side_effect).get_nullable())
+      object_factory_parameters.max_nonnull_tree_depth = 1;
+    // Get the symbol to nondet-init
+    const auto source_loc = target->source_location;
+
+    // Create aux symbol for nondet object
+    symbolt &aux_symbol = get_fresh_aux_symbol(
+      type,
+      id2string(goto_programt::get_function_id(goto_program)),
+      "nondet_tmp",
+      source_loc,
+      ID_java,
+      symbol_table);
+    aux_symbol.is_static_lifetime = false;
+
+    const symbol_exprt aux_symbol_expr = aux_symbol.symbol_expr();
+    code_declt decl(aux_symbol_expr);
+    decl.add_source_location() = source_loc;
+
+    // Create goto instructions for declaring and deading aux_symbol
+    goto_programt aux_instructions;
+    code_blockt aux_block;
+    aux_block.move(decl);
+    goto_convert(
+      aux_block, symbol_table, aux_instructions, message_handler, mode);
+    CHECK_RETURN(aux_instructions.instructions.size() == 2);
+
+    // Generate nondet init code
+    code_blockt init_code;
+
+    gen_nondet_init(
+      aux_symbol_expr,
+      init_code,
+      symbol_table,
+      source_loc,
+      true,
```

Comment:
A nice way of simulating named parameters here would be to assign this variable to a value immediately before the call:

```C++
const bool skip_classid = true
```

which of course begs the question, why is this true?

Reply:
I'm not sure - I'm copying code that was here already.